crates/egui/src/plugin.rs

```diff
@@ -0,0 +1,237 @@
+use crate::{Context, FullOutput, RawInput};
+use ahash::HashMap;
+use epaint::mutex::{Mutex, MutexGuard};
+use std::sync::Arc;
+
+/// A plugin to extend egui.
+///
+/// Add plugins via [`Context::add_plugin`].
+///
+/// Plugins should not hold a reference to the [`Context`], since this would create a cycle
+/// (which would prevent the [`Context`] from being dropped).
+#[expect(unused_variables)]
+pub trait Plugin: Send + Sync + 'static {
+    /// Plugin name.
+    ///
+    /// Used when profiling.
+    fn debug_name(&self) -> &'static str;
+
+    /// Called once, when the plugin is registered.
+    ///
+    /// Useful to e.g. register image loaders.
+    fn setup(&mut self, ctx: &Context) {}
+
+    /// Called at the start of each pass.
+    ///
+    /// Can be used to show ui, e.g. a [`crate::Window`] or [`crate::SidePanel`].
+    fn on_begin_pass(&mut self, ctx: &Context) {}
+
+    /// Called at the end of each pass.
+    ///
+    /// Can be used to show ui, e.g. a [`crate::Window`].
+    fn on_end_pass(&mut self, ctx: &Context) {}
+
+    /// Called just before the input is processed.
+    ///
+    /// Useful to inspect or modify the input.
+    /// Since this is called outside a pass, don't show ui here.
+    fn input_hook(&mut self, input: &mut RawInput) {}
+
+    /// Called just before the output is passed to the backend.
+    ///
+    /// Useful to inspect or modify the output.
+    /// Since this is called outside a pass, don't show ui here.
+    fn output_hook(&mut self, output: &mut FullOutput) {}
+}
+
+pub(crate) struct PluginHandle {
+    plugin: Box<dyn std::any::Any + Send + Sync>,
+    get_plugin_mut: fn(&mut Self) -> &mut dyn Plugin,
```

Comment:
Why not

```suggestion
    plugin: Box<dyn Plugin>,
```

?

Is it to avoid `.as_any`? But we have trait upcasting now, so that should be solvable with `Plugin: Any`… in either case: document the _why_.

Reply:
I did not use trait upcasting since we're still on 1.85. Happy to update to 1.86 though, which is the version that introduced trait upcasting.

Reply:
Ah, yeah - let's do that in a separate PR then

Reply:
Updated to use trait upcasting, much nicer now!